kaldb/src/main/java/com/slack/kaldb/server/RecoveryTaskFactory.java

```diff
@@ -0,0 +1,260 @@
+   * indexing at the offset after the recovery task.
+   *
+   * <p>When there is no offset data for a partition, return -1. In that case, the consumer would
+   * have to start indexing the data from the earliest offset.
+   */
+  public long determineStartingOffset(long currentHeadOffsetForPartition) {
+    // Filter stale snapshots for partition.
+    List<SnapshotMetadata> snapshots = snapshotMetadataStore.listSync();
+    List<SnapshotMetadata> snapshotsForPartition =
+        snapshots
+            .stream()
+            .filter(snapshotMetadata -> snapshotMetadata.partitionId.equals(partitionId))
+            .collect(Collectors.toUnmodifiableList());
+    List<SnapshotMetadata> deletedSnapshots = deleteStaleLiveSnapshots(snapshotsForPartition);
+
+    List<SnapshotMetadata> nonLiveSnapshotsForPartition =
+        snapshotsForPartition
+            .stream()
+            .filter(s -> !deletedSnapshots.contains(s))
+            .collect(Collectors.toUnmodifiableList());
+
+    // Get the highest offset that is indexed in durable store.
+    List<RecoveryTaskMetadata> recoveryTasks = recoveryTaskMetadataStore.listSync();
+    long highestDurableOffsetForPartition =
+        getHighestDurableOffsetForPartition(nonLiveSnapshotsForPartition, recoveryTasks);
+    LOG.info(
+        "The highest durable offset for partition {} is {}",
+        partitionId,
+        highestDurableOffsetForPartition);
+
+    if (highestDurableOffsetForPartition <= 0) {
+      LOG.info("There is no prior offset for this partition {}.", partitionId);
+      return highestDurableOffsetForPartition;
+    }
+
+    // The current head offset shouldn't be lower than the highest durable offset. If it is it
+    // means that we indexed more data than the current head offset. This is either a bug in the
+    // offset handling mechanism or the kafka partition has rolled over. We throw an exception
+    // for now, so we can investigate.
+    if (currentHeadOffsetForPartition < highestDurableOffsetForPartition) {
+      final String message =
+          String.format(
+              "The current head for the partition %d can't "
+                  + "be lower than the highest durable offset for that partition %d",
+              currentHeadOffsetForPartition, highestDurableOffsetForPartition);
+      LOG.error(message);
+      throw new IllegalStateException(message);
+    }
+
+    // The head offset for Kafka partition is the offset of the next message to be indexed. We
+    // assume that offset is passed into this function. The highest durable offset is the partition
+    // offset of the message that is indexed. Hence, the offset is incremented by 1 to get the
+    // next message.
+    long nextOffsetForPartition = highestDurableOffsetForPartition + 1;
+
+    // Create a recovery task if needed.
+    if (currentHeadOffsetForPartition - highestDurableOffsetForPartition > maxOffsetDelay) {
```

Comment:
We should have a minimum threshold before a recovery task will be created - if we're only lagging say a few hundred messages an indexer can easily get caught up to that. If it's a few hundred thousand we'll need a recovery task.